### ISSUE TITLE
AssetBundle $publishOptions['only'] still publishes empty directories

### ISSUE THREAD

**Issue by joe-meyer** [yiisoft/yii2#9669]
I currently have the following asset bundle:

``` php
class FontAwesomeAsset extends AssetBundle
{
    public $sourcePath = '@bower/components-font-awesome/';

    public $publishOptions = [
        'only' => [
            'css/*',
            'fonts/*',
        ],
    ];

    public $css = [
        'css/font-awesome.min.css',
    ];
}

```

Works good, but for some reason I still end up with the empty directories that aren't included in the "only" list. (less / scss). I think the "only" should apply to directories too, not just files.

![image](https://cloud.githubusercontent.com/assets/1292766/9815591/414228e6-5865-11e5-9272-a0852d36c68f.png)

**Comment by SDKiller** [comment]
I can confirm the issue

![qip shot - screen 845 21 09 15](https://cloud.githubusercontent.com/assets/2150916/9992506/58234d38-607b-11e5-8050-fa417b6046e3.png)

**Comment by quantum13** [comment]
I've written the test for this: https://github.com/quantum13/yii2/commit/6923ae31ee5735ac6ad73707be1acb9432132628

I've also found out the reason, why it does not work:
AssetManager::publishDirectory calls FileHelper::copyDirectory
FileHelper::copyDirectory calls filterPath in which option 'only' used only for files (because of e11ce60023b1da918eac8359fe674beca5c7127e)

@samdark could you please say, is it expected behaviour of option 'only' of filterPath and it shouldn't be corrected to affect to both files and dirs?

**Comment by samdark** [comment]
The commit looks intentional. I'm not sure why though... @yiisoft/core-developers ?

**Comment by SilverFire** [comment]
I guess it's not expected

**Comment by samdark** [comment]
What's not expected?

**Comment by quantum13** [comment]
I looked to commit again, it may be done for using for recursion in findFiles. If condition !is_dir will not exist, findFiles in current realization couldn't find files in directories which not satisfy "only".

**Comment by SilverFire** [comment]
It was an answer to

> @samdark could you please say, is it expected behaviour of option 'only' of filterPath and it shouldn't be corrected to affect to both files and dirs?

**Comment by quantum13** [comment]
Hello. I need help:)
There is big mismatch in docs of functions, what they realy and docs of functions which they use.
I do not know what to fix.

Chain of calls which use this $publishOptions with option 'only':
AssetBundle::publish() -> AssetManager::publish() -> AssetManager::publishDirectory() -> BaseFileHelper::copyDirectory() -> BaseFileHelper::filterPath()

AssetBundle::publish():
In guide (http://www.yiiframework.com/doc-2.0/guide-structure-assets.html#asset-options) is written that syntax 'fonts/' can be used to copy only dir fonts (i.e. whole dire fonts with subdirs and files).
It doesn't work

BaseFileHelper::copyDirectory():
In phpdoc: A path matches a pattern if it contains the pattern string at its end. For example, '.php' matches all file paths ending with '.php'.
It also doesn't work

BaseFileHelper::filterPath():
In phpdoc (lreference to BaseFileHelper::findFiles): Directory paths are not checked against them. Same pattern matching rules as in the "except" option are used (http://www.yiiframework.com/doc-2.0/yii-helpers-basefilehelper.html#findFiles()-detail)

---

Must AssetBundle::publish() and BaseFileHelper::copyDirectory() work as written in guide and doc? Or maybe docs must be rewritten?

**Comment by david-sa** [comment]
The weird behaviour of option 'only' was already reported in [#7307](https://github.com/yiisoft/yii2/issues/7307#issuecomment-140725307).

**Comment by svit-ws** [comment]
`php
        'except' => [
            'fonts/',
            'css/',
        ]
`
works fine w/o asterisks

but 'only' only with asterisks + empty folders - as mentioned above

**Comment by Bhoft** [comment]
I have the same problem...
I wanted that the less file in my assetBundle is converted but i don't want the *.less files published to the assets directory.

I tried this settings the conversion itself works perfectly without publishOptions.
But as also written here https://github.com/yiisoft/yii2/issues/2511 these 2 options were added for exactly this reason. 
So how does it work??

    public $css = [
            'css/breadcrumbs.less',
    ];

    public $publishOptions = [
            "only" => [
                "css/*",
                "js/*",
            ],
            'except' => [
                "doc/",
                //"*.less",    // css file ins't generated any more...
            ],
            "forceCopy" => YII_DEBUG,
    ];

with `"only"=>[ "css/"]` which is given in the guide http://www.yiiframework.com/doc-2.0/guide-structure-assets.html it only creates an empty css directoy. With `"only"=>[ "css/*"]`  it copies at least every file from the css directory.

The problem is with the except  *.less no css file is generated at all.

Can someone explain how the only and except parameters should work so that I get my css file generated from the less file but don't publis the less file?
I thought those parameters have been added for this issue https://github.com/yiisoft/yii2/issues/2511
